### ISSUE TITLE
splitAt traverses n elements twice

### ISSUE THREAD

**Issue by robinheghan** [elm-community/list-extra#1]
The current implementation of `splitAt` is a simple `(take n ls, drop n ls)`. This traverses `n` elements twice, first for the take, then for the drop. This could result in bad performance when `n` is a large number.

**Comment by mgold** [comment]
There's no way to avoid multiple traversals on the take part - you need to read off `n` elements in reverse order, and then reverse them. But in doing so, yes, you access the head of the drop part for free.

**Comment by jvoigtlaender** [comment]
@mgold, of course it's possible to implement this without multiple traversals.

**Comment by jvoigtlaender** [comment]
``` elm
splitAt : Int -> List a -> (List a, List a)
splitAt n list =
  if n <= 0 then
    ([], list)

  else
    case list of
      [] ->
        (list, list)

      x :: xs ->
        let (ys, zs) = splitAt (n - 1) xs
        in (x :: ys, zs)
```

**Comment by mgold** [comment]
Right.

**Comment by robinheghan** [comment]
Hmm. This could potentially lead to a stack overflow though, correct? Tradeoffs are hard :(

**Comment by jvoigtlaender** [comment]
Yes, correct. But actually it doesn't seem more likely to lead to a stack overflow than the current `core` definition for the `take`-function, and thus than the current `splitAt` definition.

**Comment by jvoigtlaender** [comment]
Since the above discussion involved the definition of `take` in `core`, and its potential for stack overflow, and since the definition of `take` in `core` just changed (https://github.com/elm-lang/core/commit/e24730e0bb687a94eaa0370d12ff3cafc098a15d), this issue here may be affected. Full discussion of the change to `take` in `core` is here: https://github.com/elm-lang/core/pull/668.

**Comment by mrbackend** [comment]
It is not possible to do this stack-safely in O(n) with one pass. To do it in one pass would require that the intermediate init was kept in order, and appending one element to a `List` is O(n), while cons-ing is O(1). So, a stack safe one pass solution would be O(n^2).

`List.take` is stack safe, at least in Elm 0.18. So, `splitAt` must be stack safe as well.

I just made a PR to `list-extra` with the main purpose to add `insertAt`. In the same PR, I also changed the implementation of `splitAt` with a slightly faster (at least on my platform) version. But it still needs two passes over the initial elements.

I move to close this issue.
